src/agent/onefuzz-agent/src/local/libfuzzer.rs

```diff
@@ -3,46 +3,101 @@
+
     let fuzzer = LibFuzzerFuzzTask::new(fuzz_config)?;
     fuzzer.check_libfuzzer().await?;
-    let fuzz_task = spawn(async move { fuzzer.run().await });
+    let mut task_handles = vec![];
+
+    let fuzz_task = spawn(async move { fuzzer.managed_run().await });
+
+    wait_for_dir(&crash_dir).await?;
 
-    let report_config = build_report_config(args)?;
-    let report = ReportTask::new(report_config);
-    let report_task = spawn(async move { report.local_run().await });
+    task_handles.push(fuzz_task);
+    if args.is_present(UNIQUE_REPORTS_DIR) {
+        let crash_report_input_monitor =
+            DirectoryMonitorQueue::start_monitoring(crash_dir.clone()).await?;
+
+        let report_config = build_report_config(
+            args,
+            Some(crash_report_input_monitor.queue_client),
+            CommonConfig {
+                task_id: Uuid::new_v4(),
+                ..common.clone()
+            },
+        )?;
+        let mut report = ReportTask::new(report_config);
+        let report_task = spawn(async move { report.managed_run().await });
+        task_handles.push(report_task);
+        task_handles.push(crash_report_input_monitor.handle);
+    }
 
     if args.is_present(COVERAGE_DIR) {
-        let coverage_config = build_coverage_config(args, true)?;
-        let coverage = CoverageTask::new(coverage_config);
-        let coverage_task = spawn(async move { coverage.local_run().await });
-
-        let result = tokio::try_join!(fuzz_task, report_task, coverage_task)?;
-        result.0?;
-        result.1?;
-        result.2?;
-    } else {
-        let result = tokio::try_join!(fuzz_task, report_task)?;
-        result.0?;
-        result.1?;
+        let coverage_input_monitor =
+            DirectoryMonitorQueue::start_monitoring(crash_dir.clone()).await?;
+        let coverage_config = build_coverage_config(
+            args,
+            true,
+            Some(coverage_input_monitor.queue_client),
+            CommonConfig {
+                task_id: Uuid::new_v4(),
+                ..common.clone()
+            },
+        )?;
+        let mut coverage = CoverageTask::new(coverage_config);
+        let coverage_task = spawn(async move { coverage.managed_run().await });
+
+        task_handles.push(coverage_task);
+        task_handles.push(coverage_input_monitor.handle);
     }
 
+    if args.is_present(ANALYZER_EXE) {
+        let analysis_input_monitor = DirectoryMonitorQueue::start_monitoring(crash_dir).await?;
+        let analysis_config = build_analysis_config(
+            args,
+            Some(analysis_input_monitor.queue_client),
+            CommonConfig {
+                task_id: Uuid::new_v4(),
+                ..common
+            },
+        )?;
+        let analysis_task = spawn(async move { run_analysis(analysis_config).await });
+
+        task_handles.push(analysis_task);
+        task_handles.push(analysis_input_monitor.handle);
+    }
+
+    try_join_all(task_handles).await?;
```

Comment:
I think this needs something equivalent to:
```suggestion
    let results = try_join_all(task_handles).await?;
    for result in results {
        result?;
    }
```
(there's definitely nicer ways to write that)

Otherwise, I think we only propagate the `JoinError` from a task that panics, not the `Ok(Err(_))` from a task that completed successfully, but whose return value was an `Err(_)`.

Reply:
(as noted elsewhere, we really need a `select()` if we want to short-circuit on the inner values)